snap/epoch.go

```diff
@@ -0,0 +1,288 @@
+		if n == 0 {
+			return ErrEpochOhSplat
+		}
+		e.Read = []uint32{n - 1, n}
+	} else {
+		e.Read = []uint32{n}
+	}
+	e.Write = []uint32{n}
+
+	return nil
+}
+
+func (e *Epoch) fromStructured(newEpoch structuredEpoch) error {
+	if newEpoch.Read == nil {
+		if newEpoch.Write == nil {
+			newEpoch.Write = []uint32{0}
+		}
+		newEpoch.Read = newEpoch.Write
+	} else if len(newEpoch.Read) == 0 {
+		// this means they explicitly set it to []. Bad they!
+		return ErrEmptyEpochList
+	}
+	if newEpoch.Write == nil {
+		newEpoch.Write = newEpoch.Read[len(newEpoch.Read)-1:]
+	} else if len(newEpoch.Write) == 0 {
+		return ErrEmptyEpochList
+	}
+	e.Read = newEpoch.Read
+	e.Write = newEpoch.Write
+
+	if err := e.Validate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *Epoch) UnmarshalJSON(bs []byte) error {
+	return e.UnmarshalYAML(func(v interface{}) error {
+		return json.Unmarshal(bs, &v)
+	})
+}
+
+func (e *Epoch) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var oldEpoch string
+	if err := unmarshal(&oldEpoch); err == nil {
+		return e.fromEpochString(oldEpoch)
+	}
+	var newEpoch structuredEpoch
+	if err := unmarshal(&newEpoch); err != nil {
+		return err
+	}
+
+	return e.fromStructured(newEpoch)
+}
+
+// Validate checks that the epoch makes sense.
+func (e *Epoch) Validate() error {
+	if e == nil || len(e.Read) == 0 || len(e.Write) == 0 {
+		return ErrEmptyEpochList
+	}
+	if !(sort.IsSorted(uint32Slice(e.Read)) && sort.IsSorted(uint32Slice(e.Write))) {
+		return ErrEpochListNotSorted
+	}
+	foundOne := false
+	for i := range e.Read {
+		idx := sort.Search(len(e.Write), func(j int) bool { return e.Write[j] >= e.Read[i] })
+		if idx < len(e.Write) && e.Write[idx] == e.Read[i] {
+			foundOne = true
+			break
+		}
+	}
+	if !foundOne {
+		return ErrNoEpochIntersection
+	}
+
+	return nil
+}
+
+func (e *Epoch) simplify() interface{} {
+	if len(e.Write) == 1 && len(e.Read) == 1 && e.Read[0] == e.Write[0] {
+		return strconv.FormatUint(uint64(e.Read[0]), 10)
+	}
+	if len(e.Write) == 1 && len(e.Read) == 2 && e.Read[0]+1 == e.Read[1] && e.Read[1] == e.Write[0] {
+		return strconv.FormatUint(uint64(e.Read[1]), 10) + "*"
+	}
+	return &structuredEpoch{Read: e.Read, Write: e.Write}
+}
+
+func (e Epoch) MarshalJSON() ([]byte, error) {
+	return json.Marshal(e.simplify())
+}
+
+func (e Epoch) String() string {
+	i := e.simplify()
+	if s, ok := i.(string); ok {
+		return s
+	}
+
+	buf, err := json.Marshal(i)
+	if err != nil {
+		// can this happen?
+		return fmt.Sprintf("%#v", e)
+	}
+	return string(buf)
+}
+
+type uint32Slice []uint32
+
+func (ns uint32Slice) Len() int           { return len(ns) }
+func (ns uint32Slice) Less(i, j int) bool { return ns[i] < ns[j] }
+func (ns uint32Slice) Swap(i, j int)      { panic("no reordering") }
+
+func parseInt(s string) (uint32, error) {
+	if len(s) > 1 && s[0] == '0' {
+		return 0, ErrBadEpochNumber
+	}
+	u, err := strconv.ParseUint(s, 10, 32)
+	if err != nil {
+		e, ok := err.(*strconv.NumError)
```

Comment:
but then `99999999999999`, which is a base 10 number with no leading zero, would fail with a message telling me it needs to be a base 10 number with no leading zeros...

Reply:
simplified the function but kept the special range case